Created blog post about integrating apps with Kubernetes (#4193)
pulumi/docs: content/blog/migrating-a-cloud-application-to-kubernetes/index.md

                                    operator: "In",
                                    values: [ebsVolume.availabilityZone],
                                }]
                            }]
                        }
                    }
                }
            }
        },
        strategy: {
            type: "Recreate"
        }
    }}, {
    deleteBeforeReplace: true,
    provider: eksCluster.provider,
});
```

In order to have a a way for the pods in cluster to communicate with our database, we create a Kubernetes service for it. It is best practice to keep databases and other similar components entirely within the cluster, and not allow them to be accessed directly the internet.

```typescript
const databasesideListener = new k8s.core.v1.Service("database-side-listener", {
    metadata: { labels: databaseDeployment.metadata.labels },
    spec: {
        type: "ClusterIP",
        ports: [{ port: 5432, targetPort: "http" }],
        selector: databaseAppLabels,
        publishNotReadyAddresses: false,
    }}, {
    provider: eksCluster.provider,
    }
);

const postgresqlAddress = databasesideListener.spec.clusterIP;
});
```

> Now, all that remains is deploying the client and server containers to Kubernetes. Since the PERN application already deployed them as docker containers, it doesn't take much effort to reconfigure it for Kubernetes. Unlike the database, these components can safely be opened up to the internet using a service with a Load Balancer type.

Reviewer: ```suggestion
Now, all that remains is deploying the client and server containers to Kubernetes. Since the PERN application already deployed them as docker containers, it doesn't take much effort to reconfigure it for Kubernetes. Unlike the database, these components can safely be open to the Internet using a service with a Load Balancer type.
```

                    env: [
                        { name: "SERVER_HOSTNAME", value: serversideListener.status.loadBalancer.ingress[0].hostname },
                    ],
                    resources: {
                        limits: {
                            memory: "500Mi",
                            cpu: "500m"
                        }
                    },
                }],
            }
        }
    }}, {
    provider: eksCluster.provider,
});

const clientsideListener = new k8s.core.v1.Service("client-side-listener", {
    metadata: { labels: clientDeployment.metadata.labels },
    spec: {
        type: "LoadBalancer",
        ports: [{ port: 3000, targetPort: "http" }],
        selector: clientAppLabels,
        publishNotReadyAddresses: false,
    }}, {
    provider: eksCluster.provider,
    }
);
```

To give us a simple way to connect to our application, we export the clientside listener's address. We can now open a browser window with the URL and port number to view our application.

```typescript
export const kubeConfig = eksCluster.kubeconfig;
export const URL = clientsideListener.status.loadBalancer.ingress[0].hostname;
```

> In this example, we demonstrated the benefits that Kubernetes offers, and showed how to use Pulumi to deploy an application to Kubernetes hosted on AWS. Although it might seem uneccessary to use such a system for only a single voting app, the advantages of running your project in the form of containerized services quickly becomes apparent as your application grows in size.

Reviewer: ```suggestion
In this example, we demonstrated the benefits Kubernetes offers and showed how to use Pulumi to deploy an application to Kubernetes hosted on AWS. Although it might seem unnecessary to use such a system for only a single voting app, the advantages of running your project in the form of containerized services become apparent as your application grows in size.
```

---
title: "Migrating a cloud application to Kubernetes"

> date: 2020-09-11

Reviewer: ```suggestion
date: 2020-09-14
```

> meta_desc: Using Pulumi to integrate applications with Kubernetes and allow for on-demand scalability of cloud architecture and freedom of design.

Reviewer: ```suggestion
meta_desc: Using Pulumi to integrate applications with Kubernetes or on-demand scalability and freedom of design.
```

There's a 60 char limit on this field

meta_image: meta.png
authors: ["vova-ivanov"]
tags: ["aws", "typescript", "docker", "kubernetes"]
---

> In this blog post, we will explore the advantages of Kubernetes, and will demonstrate how to use it by converting and deploying our [PERN application]({{< relref "/blog/deploying-a-pern-stack-application-to-aws" >}}) to Amazon EKS. With the help of Pulumi, the process becomes greatly simplified, and allows us to focus more on the big picture of designing our cloud architecture.

Reviewer 2: ```suggestion
In this blog post, we will explore the advantages of Kubernetes, and will demonstrate how to use it by converting and deploying our [PERN application]({{< relref "/blog/deploying-a-pern-stack-application-to-aws" >}}) to Amazon EKS. With the help of Pulumi, the process becomes greatly simplified and allows us to focus more on the big picture of designing our cloud architecture.
```

<!--more-->

> Kubernetes offers several distinct advantages over individually deploying Docker containers to the cloud. Firstly, is its robust system for minimizing downtime and rolling out updates. If a pod or node were to be shut down, a new one will be automatically allocated to take its place. Secondly, is the ability to scale and distribute your infrastructure. By simplifying containers into pods, Kubernetes allows you to effortlessly design the right cloud setup to fit your target. And lastly, is the interchangeability of cloud providers that comes with it. Since AWS, Azure, and GCE all support deploying Kubernetes, it means that you are not forced to stick with one provider, and have the option of easily switching if the need arises.

Reviewer 2: ```suggestion
Kubernetes offers several distinct advantages over individually deploying Docker containers to the cloud. Firstly, is its robust system for minimizing downtime and rolling out updates. If a pod or node were to be shut down, a new one will be automatically allocated to take its place. Secondly, is the ability to scale and distribute your infrastructure. By simplifying containers into pods, Kubernetes allows you to effortlessly design the right cloud setup to fit your target. And lastly, is the interchangeability of cloud providers that comes with it. Since AWS, Azure, and GCP all support deploying Kubernetes, it means that you are not forced to stick with one provider, and have the option of easily switching if the need arises.
```

```bash
$ mkdir examples && cd examples
$ git init
$ git remote add origin -f https://github.com/pulumi/examples/
$ git config core.sparseCheckout true
$ echo aws-pern-voting-app >> .git/info/sparse-checkout
$ git pull origin master
```

The next step is to create a new directory and initialize a Pulumi project with `pulumi new aws-typescript`.

```bash
$ mkdir aws-ts-k8s-voting-app && cd aws-ts-k8s-voting-app
$ pulumi new aws-typescript
```

The project will require several configuration variables, which we set using `pulumi config set`. The variables are used to configure the PostgreSQL admin account, a user account for initializing the schema and table, and the database's region.

```bash
$ pulumi config set sql-admin-name <NAME>
$ pulumi config set sql-admin-password <PASSWORD> --secret
$ pulumi config set sql-user-name <NAME>
$ pulumi config set sql-user-password <PASSWORD> --secret
$ pulumi config set aws:region <REGION>
```

The `package.json` file lists the libraries used by the project. We will add the following to the `dependencies` section:

```json
"@pulumi/eks": "^0.20.0",
"@pulumi/cloud-aws": "^0.19.0",
"@pulumi/postgresql": "^2.3.0",
"pg": "^8.3.3"
```

> Some of the components of our Kubernetified project will be almost identical to that of the PERN application's. We can copy most of these files from the example into our directory.

Reviewer 2: ```suggestion
Some of the components of our Kubernetified project will be almost identical to that of the PERN application. We can copy most of these files from the example into our directory.
```

```bash
$ cd ..
$ cp -r aws-pern-voting-app/clientside/ aws-ts-k8s-voting-app/clientside
$ cp -r aws-pern-voting-app/serverside/ aws-ts-k8s-voting-app/serverside
$ cp -r aws-pern-voting-app/postgresql_dynamic_provider.ts aws-ts-k8s-voting-app//postgresql_dynamic_provider.ts
$ cd aws-ts-k8s-voting-app
```

> The PostgreSQL Dynamic Provider used in the PERN app to create schemas will need to be slightly modified to include an `ssl: false` pool parameter in the `create` and `delete` functions.

Reviewer 2: You might expand upon this to describe why you need to do this.

        const
        scriptExecuted = await pool.query(args.deletionScript);
        await pool.end();
    }

    async diff(id: string, oldArgs: SchemaInputs, newArgs: SchemaInputs): Promise<pulumi.dynamic.DiffResult> {...}
    async update(id: string, oldArgs: SchemaInputs, newArgs: SchemaInputs): Promise<pulumi.dynamic.UpdateResult> {...}
}
```

Likewise, we need to add an `ssl: false` option to `serverside/server/db.js` as well.

```typescript
const Pool = require("pg").Pool;

const pool = new Pool({
    user: process.env["USER_NAME"],
    password: process.env["USER_PASSWORD"],
    host: process.env["POSTGRES_ADDRESS"],
    port: process.env["POSTGRES_PORT"],
    database: process.env["DATABASE_NAME"],
    ssl: false,
});

module.exports = pool;
```

In order to fully prepare our application for Kubernetes, a Docker container has to be created to act as the PostgreSQL database for the application. Although this is slightly more complex than directly asking AWS to provision an RDS instance, the underlying mechanics behind both approaches are the same.

The first step, is to create two directories:

```bash
$ mkdir databaseside && cd databaseside
$ mkdir database
```

> Similarily to the PERN app, the outer `databaseside` folder will hold our Dockerfile, while the inner `database` folder will contain any code that we need.

Reviewer 2: ```suggestion
Similar to the PERN app, the outer `databaseside` folder will hold our Dockerfile, while the inner `database` folder will contain any code that we need.
```

const sqlAdminPassword = config.requireSecret("sql-admin-password");
const sqlUserName = config.require("sql-user-name");
const sqlUserPassword = config.requireSecret("sql-user-password");
const awsConfig = new pulumi.Config("aws");
const region = aws.config.region;
```

The central part of our application will be an AWS Elastic Kubernetes Cluster. It holds our application, and dictates how many nodes should be allocated for the processes running inside the cluster.

```typescript
const eksCluster = new eks.Cluster("eksCluster", {
    name: "eksCluster",
    instanceType: "t2.medium",
    desiredCapacity: 3,
    minSize: 2,
    maxSize: 4,
    deployDashboard: false,
    enabledClusterLogTypes: [
        "api",
        "audit",
        "authenticator",
    ],
});
```

Kubernetes nodes use ephemeral storage, meaning that when they shut down or get restarted, everything is erased. To create a permanent storage system suitable for our database, we need a Persistent Volume instance.

```typescript
const ebsVolume = new aws.ebs.Volume("storage-volume", {
    size: 10,
    type: "gp2",
    availabilityZone: region + "a",
    encrypted: true,
    }, {
    protect: false,
});
```

> Now that the infrastructure for our database is ready, we can launch our very own PostgreSQL server on the cloud in the form of a Deployment. Deployments offer a multuitude of different customization options ranging from attaching volumes, setting the memory and cpu requirements, to specifying which deployments can and cannot be run on the same node.

Reviewer 2: ```suggestion
Now that the infrastructure for our database is ready, we can launch our very own PostgreSQL server on the cloud in the form of a Deployment. Deployments offer a multitude of different customization options ranging from attaching volumes, setting the memory and CPU requirements, to specifying which deployments can and cannot be run on the same node.
```

                affinity: {  
                    nodeAffinity: {
                        requiredDuringSchedulingIgnoredDuringExecution: {
                            nodeSelectorTerms: [{
                                matchExpressions: [{
                                    key: "failure-domain.beta.kubernetes.io/zone",
                                    operator: "In",
                                    values: [ebsVolume.availabilityZone],
                                }]
                            }]
                        }
                    }
                }
            }
        },
        strategy: {
            type: "Recreate"
        }
    }}, {
    deleteBeforeReplace: true,
    provider: eksCluster.provider,
});
```

In order to have a clean, simple URL through which we can access the database, we create a Kubernetes service for it.

```typescript
const databasesideListener = new k8s.core.v1.Service("database-side-listener", {
    metadata: { labels: databaseAppLabels },
    spec: {
        type: "LoadBalancer",
        ports: [{ port: 5432, targetPort: "http" }],
        selector: databaseAppLabels,
    }}, {
    provider: eksCluster.provider,
    }
);

> 

Reviewer: This needs to be changed in the project. A [load balancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) would expose PGSQL outside of the cluster, which is bad practice. Connection to the database should only use clusterIPs, if the application is in the same Pod/Node you can use selectors.

Author: It's definitely true that exposing PGSQL to the internet is bad practice, but the Pulumi PostgreSQL provider that I used seems to require a direct server address to connect to. If no external address is given to the database, then I don't think the provider is able to function.

Author: Removed the providers entirely, and instead changed the project to perform the initialization steps in the container bash script. The database now uses a ClusterIP service, and is only accessible from within the cluster.

const postgresqlAddress = databasesideListener.status.loadBalancer.ingress[0].hostname;
});
```

> We can now treat our Kubernetes database the same way as we did our RDS one. Like the PERN application, we create a user, schema, and table, grant permissions for our user to edit and select it, and populate the table with two initial voting options.

Reviewer: ```suggestion
We can use ourt Postgres database in the same way as our RDS version. Like the PERN application, we create a user, schema, and table, grant permissions for our user to edit and select it, and populate the table with two initial voting options.
```

    name: "votes"}, {
    provider: postgresqlProvider,
});

const postgresUser = new postgresql.Role("postgres-standard-role", {
    name: sqlUserName,
    password: sqlUserPassword,
    superuser: false,
    login: true,
    connectionLimit: -1}, {
    provider: postgresqlProvider,
});

const creationScript = `
    CREATE SCHEMA voting_app;
    CREATE TABLE voting_app.choice(
        choice_id SERIAL PRIMARY KEY,
        text VARCHAR(255) NOT NULL,
        vote_count INTEGER NOT NULL
    );
    GRANT USAGE ON SCHEMA voting_app TO ${sqlUserName};
    GRANT SELECT, UPDATE ON ALL TABLES IN SCHEMA voting_app TO ${sqlUserName};
    INSERT INTO voting_app.choice (text, vote_count) VALUES('Tabs', 0);
    INSERT INTO voting_app.choice (text, vote_count) VALUES('Spaces', 0);
`;

const deletionScript = "DROP SCHEMA IF EXISTS voting_app CASCADE";

const postgresqlVotesSchema = new Schema("postgresql-votes-schema", {
    creatorName: sqlAdminName,
    creatorPassword: sqlAdminPassword,
    serverAddress: postgresqlAddress,
    databaseName: postgresDatabase.name,
    creationScript: creationScript,
    deletionScript: deletionScript,
    postgresUserName: postgresUser.name,
});
```

> And now, all that remains is the deploy the client and server containers to Kubernetes. Since the PERN application already deplyed them as docker containers, it doesn't take too much effort to reconfigure it for Kubernetes.

Reviewer 2: ```suggestion
And now, all that remains is to deploy the client and server containers to Kubernetes. Since the PERN application already deployed them as docker containers, it doesn't take too much effort to reconfigure it for Kubernetes.
```

> And now, all that remains is the deploy the client and server containers to Kubernetes. Since the PERN application already deplyed them as docker containers, it doesn't take too much effort to reconfigure it for Kubernetes.

Reviewer: ```suggestion
And now, all that remains is deploying the client and server containers to Kubernetes. Since the PERN application already deplyed them as docker containers, it doesn't take much effort to reconfigure it for Kubernetes.
```

                    ports: [{ name: "http", containerPort: 3000 }],
                    env: [
                        { name: "SERVER_HOSTNAME", value: serversideListener.status.loadBalancer.ingress[0].hostname },
                    ],
                    resources: {
                        limits: {
                            memory: "500Mi",
                            cpu: "500m"
                        }
                    },
                }],
            }
        }
    }}, {
    provider: eksCluster.provider,
});

const clientsideListener = new k8s.core.v1.Service("client-side-listener", {
    metadata: { labels: clientAppLabels },
    spec: {
        type: "LoadBalancer",
        ports: [{ port: 3000, targetPort: "http" }],
        selector: clientAppLabels,
    }}, {
    provider: eksCluster.provider,
    }
);
```

To make our Kubernetes application available on the Internet, we export the clientside listener's address. We can open a browser window with the URL and port to view our application.

```typescript
export const URL = clientsideListener.status.loadBalancer.ingress[0].hostname;
```

> In this example, we went over the benefits that Kubernetes offers, and showed how to use Pulumi to convert an application to Kubernetes and deploy it on AWS. Although it might seem uneccessary to use such a system for only a single voting app, the advantages of running your project in the form of containerized services quickly become apprant as your application grows in size.  

Reviewer 2: ```suggestion
In this example, we went over the benefits that Kubernetes offers and showed how to use Pulumi to convert an application to Kubernetes and deploy it on AWS. Although it might seem unnecessary to use such a system for only a single voting app, the advantages of running your project in the form of containerized services quickly become apparent as your application grows in size.  
```

> In this example, we went over the benefits that Kubernetes offers, and showed how to use Pulumi to convert an application to Kubernetes and deploy it on AWS. Although it might seem uneccessary to use such a system for only a single voting app, the advantages of running your project in the form of containerized services quickly become apprant as your application grows in size.  

Reviewer: ```suggestion
In this example, we demonstrated the benefits that Kubernetes offers, and showed how to use Pulumi to deploy an application to Kubernetes hosted on AWS. Although it might seem uneccessary to use such a system for only a single voting app, the advantages of running your project in the form of containerized services quickly becomes apparent as your application grows in size.  
```

> Next week, I'll ____.

Reviewer 2: Perhaps look at the MERN stack next week (MongoDB is the 'M')
